### PR TITLE
ts_project ESM support

### PR DESCRIPTION
**Pull request by Toxicable**
## PR Checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)

## What is the current behavior?
Issue Number: #3277

## What is the new behavior?
Adds ESM support in for  `ts_project`

## Does this PR introduce a breaking change?
- [x] No - Only changes are internal API's

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -55,7 +55,7 @@ load("@rules_nodejs//nodejs:repositories.bzl", "nodejs_register_toolchains")
 # This applies to the resolved_toolchain
 nodejs_register_toolchains(
     name = "node16",
-    node_version = "16.5.0",
```
Comment (Toxicable) [inline WORKSPACE:58]: Note that this change was required to for @angular/core - they require nodejs >=16.10.0

**Comment by diceride** [comment]
@Toxicable Thank you working on this, I'm also strugling with the same problem.

Your implementation introduces on a new .mts file extension. The fact that Node.js relies on a .mjs file extension to detect whether to use JavaScript module syntax is a unique feature of Node.js only. In my opinion it's a bad design choice by the Node.js team. By introducing a new .mts file extension, Node.js behaviour would spill over into the TypeScript world. I'm not sure whether this is the right approach to resolve the issue.

**Comment by diceride** [comment]
@Toxicable Thank you working on this, I'm also strugling with the same problem.

Your implementation introduces on a new .mts file extension. The fact that Node.js relies on a .mjs file extension to detect whether to use JavaScript module syntax is a unique feature of Node.js only. In my opinion it's a bad design choice by the Node.js team. By introducing a new .mts file extension, Node.js behaviour would spill over into the TypeScript world. I'm not sure whether this is the right approach to resolve the issue. File extensions are ignored by web browsers and by Deno too.

**Comment by Toxicable** [comment]
>Your implementation introduces on a new .mts file extension

@diceride Unfortunately the ship has sailed and this is the deisgn that has been chosen by both NodeJS and the TypeScript Teams. I'm not keen to introduce any new designs here and would prefer to stick to what has already decided on.
The goal of this PR is to support the same set of file extensions that are already in TypeScript as outlined here https://github.com/bazelbuild/rules_nodejs/pull/3298 
In saying that, there's still the option to use the `package.json#type=module` method if that helps

**Comment by alexeagle** [comment]
CI failure
```
error TS18003: No inputs were found in config file 'C:/b/makknmlb/execroot/build_bazel_rules_nodejs/packages/typescript/test/ts_project/esm/tsconfig.json'. Specified 'include' paths were '["**/*"]' and 'exclude' paths were '[]'.
```
is probably because of Windows lack of sandboxing by default - do you need help to make it green?

**Comment by Toxicable** [comment]
@alexeagle yeah that would be awesome if you got the time

**Comment by alexeagle** [comment]
Note: https://github.com/bazelbuild/rules_nodejs/pull/3405 seems like the prerequisite but needs someone with some time to green it up...

**Pull request closed without merging**